### PR TITLE
Allow VIRTUAL_PATH routing (sub-VIRTUAL_HOST)

### PR DESCRIPTION
**Pull request by trowj**
This is really an update to the fork at 'https://github.com/gregsymons/docker-nginx-proxy' that I believe I fully updated to current master and merged the functionality.  I will honestly say that I can't get the tests to run  due to some kind of python bug I think (I can't run them on this current repo either so it's not a change I made at least) so I cannot be sure everything is still passing, but I did my best in comparison locally between versions as to what was generated for default.conf

### CODE REVIEW

**Comment by Mklueh** [comment]
I really hope this get´s included in the main image as I´m trying to run this whole thing on my private server (no domain, only IP) - seems to be impossible without VIRTUAL_PATH. See #883

No clue how I can get my application running without that. The only option I see is setting up and configuring an nginx proxy manually

**Comment by trowj** [comment]
@Mklueh You can pull my customized version from 'joshtrow/nginx-proxy', it has this PR in it for now. I'm trying to figure out the tests now (why I can't run them, and what to do to fix the issues noted above)

**Comment by rodrigoaguilera** [comment]
I am also interested in this feature.

I'm trying to review this PR but I noticed something odd.
There some commits by Greg  Symons listed but they don't seem to be listed as "Files changed"
Why is this strange behavior?

File: nginx.tmpl [hunk 1]
```diff
@@ -17,6 +17,59 @@
...
+		include /etc/nginx/vhost.d/default_location;
+	{{ end }}
+}
+{{ end }}
+
+{{ define "upstream-definition" }}
+	upstream {{ .Upstream }} {
+		{{ range $container := .Containers }}
+			{{ $addrLen := len $container.Addresses }}
+			{{ $networks := .Networks }}
```
Comment (bibendi) [inline nginx.tmpl]: https://github.com/abak-press/nginx-proxy/pull/2

**Comment by bibendi** [comment]
https://hub.docker.com/r/abakpress/nginx-proxy/

**Comment by rodrigoaguilera** [comment]
At one of the commits you mention

>  Removing trailing dollar signs due to nginx blowing up - and rather than go whole-hog on this as "appropriate@89bff71" seems to do, just keep the commit to one fix at a time

Why not sha1 the upstream name when is a regex like this repo (jwilder/nginx-proxy) does? 
appropriate/nginx-proxy seems like a very old fork to be an example.

I think also you are missing a trim of the of host name.

**Comment by rodrigoaguilera** [comment]
I did a pull request to check if that was the problem of the failing tests and it seems that is the issue.

https://github.com/jwilder/nginx-proxy/pull/1083

This is the corresponding branch that also has a rebase against the current master

https://github.com/upstreamable/nginx-proxy/tree/virtual_path

Feel free to take any work from that branch and incorporate it in this one.

@bibendi Can you explain more about your review even if simple words or auto translated?

**Comment by bibendi** [comment]
We had 3 problems:

1) https://github.com/abak-press/nginx-proxy/pull/1#discussion-diff-167620140R30
With trailing slash of a location service container gave a trimmed URI.

2) https://github.com/abak-press/nginx-proxy/pull/1#discussion-diff-167620328R170
Allow to have multiple paths for one service container.

3) https://github.com/abak-press/nginx-proxy/pull/2/files#diff-6c19a779564e1e2f88854a4a2fffbb78R47
Fixed a strange bag with wrong generated networks.

**Comment by rodrigoaguilera** [comment]
@bibendi Thank you!

I incorporated all your fixes into https://github.com/jwilder/nginx-proxy/pull/1083

Except for the multiple VIRTUAL_PATH separated by commas. I feel that can be a feature once this one get in.

**Comment by neilst** [comment]
Any update on this getting merged into master ?

File: nginx.tmpl [hunk 2]
```diff
@@ -112,36 +167,20 @@ server {
...
+{{ $paths := groupBy $containers "Env.VIRTUAL_PATH" }}
+{{ $nPaths := len $paths }}
+
+{{ if eq $nPaths 0 }}
+	# {{ $host }}
+	{{ template "upstream-definition" (dict "Upstream" $host "Containers" $containers "Networks" (json $CurrentContainer.Networks)) }}
+{{ else }}
+	{{ range $path, $containers := $paths }}
+		{{ $sum := sha1 $path }}
+		{{ $upstream := printf "%s-%s" $host $sum }}
```
Comment (kfkonrad) [inline nginx.tmpl]: Without this line, I get a "too many redirects" errors. So I get that this has to be there, but I don't understand why or how this works. I also found nothing in the docs, can you elaborate?

**Comment by buchdag** [comment]
This features has been added to the `dev` branch with #1607 

It's available now on the `dev` images `nginxproxy/nginx-proxy:dev` and `nginxproxy/nginx-proxy:dev-alpine`. Testing and feedback are welcome, the docs for the feature is [here](https://github.com/nginx-proxy/nginx-proxy/tree/dev#path-based-routing) and there is [another PR](https://github.com/nginx-proxy/nginx-proxy/pull/1699) opened for additional improvements and comments.

**Pull request closed without merging**